### PR TITLE
issue-469, delete all users from deleted cluster

### PR DESCRIPTION
**Pull request by OleksiienkoMykyta**

### CODE REVIEW

**Review by worryg0d — APPROVED** [review]

**Review by ribaraka — CHANGES_REQUESTED** [review]

File: controllers/clusterresources/redisuser_controller.go [hunk 1]
```diff
@@ -213,10 +213,23 @@ func (r *RedisUserReconciler) Reconcile(ctx context.Context, req ctrl.Request) (
...
 		}
 
+		if event == models.ClusterDeletingEvent {
+			err = r.detachUserFromDeletedCluster(ctx, clusterID, user, l)
+			if err != nil {
+				l.Error(err, "Cannot delete Redis user resource",
+					"secret name", secret.Name,
+					"secret namespace", secret.Namespace)
+				r.EventRecorder.Eventf(user, models.Warning, models.PatchFailed,
+					"Resource deleting is failed. Reason: %v", err)
```
[the remark this answers is not in the document]
Comment (OleksiienkoMykyta) [reply controllers/clusterresources/redisuser_controller.go]: Fixed

File: controllers/clusterresources/redisuser_controller.go [hunk 2]
```diff
@@ -254,6 +267,56 @@ func (r *RedisUserReconciler) Reconcile(ctx context.Context, req ctrl.Request) (
...
+		)
+		r.EventRecorder.Eventf(
+			user, models.Warning, models.PatchFailed,
+			"Detaching clusterID from the Redis user resource has been failed. Reason: %v",
+			err,
+		)
+		return err
+	}
+
+	patch = user.NewPatch()
```
[the remark this answers is not in the document]
Comment (OleksiienkoMykyta) [reply controllers/clusterresources/redisuser_controller.go]: Fixed

File: controllers/clusters/redis_controller.go [hunk 3]
```diff
@@ -476,6 +477,23 @@ func (r *RedisReconciler) handleDeleteCluster(
 		return models.ReconcileRequeue
 	}
 
+	for _, ref := range redis.Spec.UserRefs {
+		err = r.detachUserResource(ctx, logger, redis, ref)
+		if err != nil {
+			logger.Error(err, "Cannot delete Redis user cluster",
```
[the remark this answers is not in the document]
Comment (OleksiienkoMykyta) [reply controllers/clusters/redis_controller.go]: Fifed

File: controllers/clusterresources/redisuser_controller.go [hunk 4]
```diff
@@ -139,28 +139,28 @@ func (r *RedisUserReconciler) Reconcile(ctx context.Context, req ctrl.Request) (
...
 			if finalizerNeeded {
 				err = r.Update(ctx, secret)
 				if err != nil {
-					l.Error(err, "Cannot update Cassandra user secret",
+					l.Error(err, "Cannot update Redis user secret with deletion finalizer",
 						"secret name", secret.Name,
 						"secret namespace", secret.Namespace)
 					r.EventRecorder.Eventf(user, models.Warning, models.UpdatedEvent,
-						"Cannot assign Cassandra user to a k8s secret. Reason: %v", err)
+						"Cannot assign Redis user to a k8s secret. Reason: %v", err)
```
Comment (testisnullus) [inline controllers/clusterresources/redisuser_controller.go:150]: `Cannot assign k8s secret to a Redis user` maybe?

File: controllers/clusterresources/redisuser_controller.go [hunk 5]
```diff
@@ -174,9 +174,9 @@ func (r *RedisUserReconciler) Reconcile(ctx context.Context, req ctrl.Request) (
 			userID := fmt.Sprintf(instaclustr.RedisUserIDFmt, clusterID, user.Namespace)
 			err = r.API.DeleteRedisUser(userID)
 			if err != nil {
-				l.Error(err, "Cannot delete Redis user", "user", user.Name)
+				l.Error(err, "Cannot delete Redis user on Instaclustr", "user", user.Name)
```
Comment (testisnullus) [inline controllers/clusterresources/redisuser_controller.go:177]: What does `Cannot delete Redis user on Instaclustr` mean? We delete users from a cluster, isn't it?

File: controllers/clusterresources/redisuser_controller.go [hunk 6]
```diff
@@ -254,6 +267,55 @@ func (r *RedisUserReconciler) Reconcile(ctx context.Context, req ctrl.Request) (
...
+		)
+		r.EventRecorder.Eventf(
+			user, models.Warning, models.PatchFailed,
+			"Detaching clusterID from the Redis user resource has been failed. Reason: %v",
+			err,
+		)
+		return err
+	}
+
+	controllerutil.RemoveFinalizer(user, user.DeletionUserFinalizer(clusterID, user.Namespace))
```
Comment (testisnullus) [inline controllers/clusterresources/redisuser_controller.go:293]: `DeletionUserFinalizer` -> `DeletionFinalizer`

File: controllers/clusters/redis_controller.go [hunk 7]
```diff
@@ -476,6 +477,23 @@ func (r *RedisReconciler) handleDeleteCluster(
 		return models.ReconcileRequeue
 	}
 
+	for _, ref := range redis.Spec.UserRefs {
+		err = r.detachUserResource(ctx, logger, redis, ref)
+		if err != nil {
+			logger.Error(err, "Cannot detach Redis user cluster",
```
Comment (testisnullus) [inline controllers/clusters/redis_controller.go:483]: `Cannot detach Redis user cluster` -> `Cannot detach Redis user`

**Review by testisnullus — CHANGES_REQUESTED** [review]

File: controllers/clusterresources/redisuser_controller.go (hunk 4 above)
Comment (OleksiienkoMykyta) [reply controllers/clusterresources/redisuser_controller.go]: Yes, you are right, thank you!
Fixed

File: controllers/clusterresources/redisuser_controller.go (hunk 5 above)
Comment (OleksiienkoMykyta) [reply controllers/clusterresources/redisuser_controller.go]: Yes, fixed

File: controllers/clusterresources/redisuser_controller.go (hunk 6 above)
Comment (OleksiienkoMykyta) [reply controllers/clusterresources/redisuser_controller.go:293]: It's a separate method for user,  DeletionFinalizer used for secrets

File: controllers/clusters/redis_controller.go (hunk 7 above)
Comment (OleksiienkoMykyta) [reply controllers/clusters/redis_controller.go]: fixed

File: controllers/clusterresources/redisuser_controller.go [hunk 8]
```diff
@@ -174,9 +174,9 @@ func (r *RedisUserReconciler) Reconcile(ctx context.Context, req ctrl.Request) (
 			userID := fmt.Sprintf(instaclustr.RedisUserIDFmt, clusterID, user.Namespace)
 			err = r.API.DeleteRedisUser(userID)
 			if err != nil {
-				l.Error(err, "Cannot delete Redis user", "user", user.Name)
+				l.Error(err, "Cannot delete Redis user on cluster. Cluster ID: %s.", "user", clusterID, user.Name)
```
Comment (testisnullus) [inline controllers/clusterresources/redisuser_controller.go:177]: Cannot delete Redis user on cluster -> Cannot delete Redis user from the cluster

**Review by testisnullus — CHANGES_REQUESTED** [review]

File: controllers/clusters/redis_controller.go [hunk 9]
```diff
@@ -596,6 +614,52 @@ func (r *RedisReconciler) handleDeleteCluster(
...
+			l.Error(err, "Redis user is not found", "request", req)
+			r.EventRecorder.Eventf(redis, models.Warning, models.NotFound,
+				"User resource is not found, please provide correct userRef."+
+					"Current provided reference: %v", uRef)
+			return err
+		}
+
+		l.Error(err, "Cannot get Redis user", "user", u.Spec)
+		r.EventRecorder.Eventf(redis, models.Warning, models.DeletionFailed,
+			"Cannot get Redis user. user reference: %v", uRef)
```
[the remark this answers is not in the document]
Comment (OleksiienkoMykyta) [reply controllers/clusters/redis_controller.go]: Fixed

**Review by taaraora — APPROVED** [review]

**Pull request merged**